modules/basicfuncs/list-funcs.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2016 Balabit
+ * Copyright (c) 2016 Balázs Scheidler
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 as published
+ * by the Free Software Foundation, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "scanner/list-scanner/list-scanner.h"
+#include "str-repr/encode.h"
+
+static void
+_append_comma(GString *result)
+{
+  if (result->len == 0)
+    return;
+
+  if (result->str[result->len - 1] != ',')
+    g_string_append_c(result, ',');
+}
+
+static void
+tf_list_concat(LogMessage *msg, gint argc, GString *argv[], GString *result)
+{
+  ListScanner scanner;
+
+  list_scanner_init(&scanner);
+  list_scanner_input_gstring_array(&scanner, argc, argv);
+  while (list_scanner_scan_next(&scanner))
+    {
+      _append_comma(result);
+      str_repr_encode_append(result, list_scanner_get_current_value(&scanner), -1, ",");
+    }
+  list_scanner_deinit(&scanner);
+}
+
+TEMPLATE_FUNCTION_SIMPLE(tf_list_concat);
+
+static void
+tf_list_append(LogMessage *msg, gint argc, GString *argv[], GString *result)
+{
+  if (argc == 0)
+    return;
+  g_string_append_len(result, argv[0]->str, argv[0]->len);
+  for (gint i = 1; i < argc; i++)
+    {
+      _append_comma(result);
+      str_repr_encode_append(result, argv[i]->str, argv[i]->len, ",");
+    }
+}
+
+TEMPLATE_FUNCTION_SIMPLE(tf_list_append);
+
+static gint
+_list_count(gint argc, GString *argv[])
+{
+  ListScanner scanner;
+  gint count = 0;
+
+  if (argc != 0)
+    {
+      list_scanner_init(&scanner);
+      list_scanner_input_gstring_array(&scanner, argc, argv);
+
+      while (list_scanner_scan_next(&scanner))
+        count++;
+
+      list_scanner_deinit(&scanner);
+    }
+  return count;
+}
+
+static void
+_adjust_list_slice(gint argc, GString *argv[],
+                   gint *first_ndx, gint *last_ndx)
+{
+  gint count = -1;
+
+  if (*first_ndx < 0 || *last_ndx < 0)
+    count = _list_count(argc, argv);
+
+  if (*first_ndx < 0)
+    *first_ndx += count;
```

Comment:
parse_number() can parse negative numbers, it does parsing with strtoll(). 

The negative values are actually supported values in slice boundaries and are interpreted similarly to Python does that.

```
$(list-slice -1: $LIST) - is the last element (negative elements index from the end of the list)
$(list-slice :-1 $LIST) - is until the last element
```

_adjust only adds "count", which might still be negative, but as I see those cases are handled in the caller. as the loop to skip initial entries would not be entered, thus we would be using entries from the very first element.

```
$(list-slice -100: a,b,c) - returns "a,b,c"
$(list-slice -1: a,b,c) - returns "c"
```
Similarly for the end index. If it is too low, the 'slicing' loop will not be entered (the 2nd loop in _list_slice), causing no elements of the list being returned.

Anyway, adding a comment to make this clear.